bot/wikidata/federation_tester.py

```diff
@@ -46,49 +51,52 @@ def run(self):
+        newtext += '{| class="wikitable sortable"\n! Link !! Git !! Wiki !! Query works?\n'
         alllist = sorted(set(self.gitlist) | set(self.wikilist))
         for link in alllist:
-            newtext += u'|-\n'
-            newtext += u'| %s\n' % (link,)
+            newtext += '|-\n'
+            newtext += '| %s\n' % (link,)
             if link in self.gitlist:
-                newtext += u'| OK\n' # FIXME: Add pretty icon
+                newtext += '| OK\n' # FIXME: Add pretty icon
             else:
-                newtext += u'| Missing\n'
+                newtext += '| Missing\n'
             if link in self.wikilist:
-                newtext += u'| OK\n' # FIXME: Add pretty icon
+                newtext += '| OK\n' # FIXME: Add pretty icon
             else:
-                newtext += u'| Missing\n'
+                newtext += '| Missing\n'
             queryworks = self.testSparql(link)
             if queryworks:
-                newtext += u'| OK\n' # FIXME: Add pretty icon
+                newtext += '| OK\n' # FIXME: Add pretty icon
             else:
-                newtext += u'| Query failed\n'
-        newtext += u'|}\n[[Category:Wikidata:SPARQL query service]]'
-        print newtext
-        reportpage.put(newtext, summary=u'Updating federation report')
+                newtext += '| Query failed\n'
+        newtext += '|}\n[[Category:Wikidata:SPARQL query service]]'
+        print(newtext)
+        reportpage.put(newtext, summary='Updating federation report')
 
     def testSparql(self, link):
-        query = u"""SELECT ?a ?b ?c WHERE { 
+        query = """SELECT ?a ?b ?c WHERE { 
   SERVICE <%s> {
     SELECT ?a ?b ?c WHERE { ?a ?b ?c } LIMIT 1
   }
 } LIMIT 1""" % (link,)
+        headers = pywikibot.data.sparql.DEFAULT_HEADERS | {
+            'User-Agent': 'federation_tester.py (multichill)',
+        }
         sq = pywikibot.data.sparql.SparqlQuery(max_retries=1, retry_wait=1)
         try:
-            queryresult = sq.select(query)
+            queryresult = sq.select(query, headers=headers)
             if queryresult:
                 return True
-            print queryresult
+            print(queryresult)
             #for resultitem in queryresult:
-            #    qid = resultitem.get('item').replace(u'http://www.wikidata.org/entity/', u'')
+            #    qid = resultitem.get('item').replace('http://www.wikidata.org/entity/', u'')
             #    result[resultitem.get('id')] = qid
-        except pywikibot.exceptions.TimeoutError:
+        except (pywikibot.exceptions.TimeoutError, pywikibot.exceptions.ServerError):
```

Comment:
Exceptions for pywikibot have changed a lot in the last years, so probably wise to double check if these are still correct.

Reply:
I added a `print(type(e))` and it appears that both `TimeoutError` and `ServerError` can be raised for different endpoints.